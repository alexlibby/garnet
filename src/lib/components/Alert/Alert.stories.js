import Alert from "./Alert.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Notification Components/Alert",
  component: Alert,
  argTypes: {
    type: { control: "string" },
    title: { control: "string" },
    description: { control: "string" },
    showIcon: { action: "boolean" },
  },
  parameters: {
    docs: {
      story: {
        height: "100px",
      },
    },
  },
};

export const Default = () => ({
  Component: Alert,
  props: {
    show: true,
    type: "info",
    title: "Simple Info",
    description: "An info description",
    showIcon: "true",
  },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const Warn = () => ({
  Component: Alert,
  props: {
    show: true,
    type: "warn",
    title: "Simple warning",
    description: "A warning message",
    showIcon: "true",
  },
});

Warn.parameters = {
  badges: [BADGES.ALPHA],
};

export const Error = () => ({
  Component: Alert,
  props: {
    show: true,
    type: "error",
    title: "Error message",
    description: "An error message",
    showIcon: "true",
  },
});

Error.parameters = {
  badges: [BADGES.ALPHA],
};
