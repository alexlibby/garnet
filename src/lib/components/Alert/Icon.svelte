<svelte:options customElement="garnet-icon" />

<script>
  export let width = "24px";
  export let height = "24px";
  export let iconType = "";

  let icons = [
    {
      box: 512,
      name: "warn",
      svg: `<path d="M85.57 446.25h340.86a32 32 0 0028.17-47.17L284.18 82.58c-12.09-22.44-44.27-22.44-56.36 0L57.4 399.08a32 32 0 0028.17 47.17z" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="32"/><path d="M250.26 195.39l5.74 122 5.73-121.95a5.74 5.74 0 00-5.79-6h0a5.74 5.74 0 00-5.68 5.95z" fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="32"/>
  <path d="M256 397.25a20 20 0 1120-20 20 20 0 01-20 20z"/>`,
    },
    {
      box: 512,
      name: "info",
      svg: `<path d="M248 64C146.39 64 64 146.39 64 248s82.39 184 184 184 184-82.39 184-184S349.61 64 248 64z" fill="none" stroke="currentColor" stroke-miterlimit="10" stroke-width="32"/><path fill="none" stroke="#ffffff" stroke-linecap="round" stroke-linejoin="round" stroke-width="32" d="M220 220h32v116"/>
      <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="32" d="M208 340h88" /><path d="M248 130a26 26 0 1026 26 26 26 0 00-26-26z" fill="#ffffff" />`,
    },
    {
      box: 512,
      name: "error",
      svg: `<path d="M448 256c0-106-86-192-192-192S64 150 64 256s86 192 192 192 192-86 192-192z" stroke="currentColor" fill="none" stroke-miterlimit="10" stroke-width="32"/>
      <path d="M250.26 166.05L256 288l5.73-121.95a5.74 5.74 0 00-5.79-6h0a5.74 5.74 0 00-5.68 6z" fill="#ffffff" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="32"/><path d="M256 367.91a20 20 0 1120-20 20 20 0 01-20 20z" fill="#ffffff" />`,
    },
  ];

  let displayIcon = icons.find((e) => e.name === iconType);
</script>

<svg
  class={$$props.class}
  {width}
  {height}
  viewBox="0 0 {displayIcon.box} {displayIcon.box}"
>
  {@html displayIcon.svg}
</svg>
