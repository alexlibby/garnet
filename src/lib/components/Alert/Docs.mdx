import { Canvas, Meta, Story } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as AlertStories from "./Alert.stories";

<Meta of={AlertStories} />

# Alert

Alert is the primary component. It has three possible states.

- [Default](#default)
- [Warn](#warn)
- [Error](#error)

## Default

This is the default state for the Alert component - it shows an information message:

<Canvas of={AlertStories.Default} />

## Warn

This is the warning state for the Alert component - it shows a warning message:

<Canvas of={AlertStories.Warn} />

## Error

This is the error state for the Alert component - it shows an error message:

<Canvas of={AlertStories.Error} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={AlertStories} />
