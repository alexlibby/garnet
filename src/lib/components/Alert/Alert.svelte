<svelte:options customElement="garnet-alert" />

<script>
  import Icon from "./Icon.svelte";

  export let show;
  export let showIcon = true;
  export let type = "";
  export let title = "";
  export let description = "";
  export let showAnimation = true;
  let typeClass;

  switch (type) {
    case "warn":
      typeClass = "alert-warn";
      break;
    case "dark":
      typeClass = "alert-dark";
      break;
    case "error":
      typeClass = "alert-error";
      break;
    case "info":
      typeClass = "alert-info";
      break;
    case "success":
      typeClass = "alert-success";
      break;
    default:
      typeClass = "";
  }
  const classes = ["alert", typeClass, showAnimation ? "fade-in" : ""]
    .join(" ");

  const closeAlert = () => {
    show = false;
  };
</script>

<div class="garnet-alert">
  {#if show}
    <dialog class={classes} role="alertdialog" title="Alert Dialog">
      <div class="icon">
        {#if showIcon}<Icon iconType={type} />{/if}
      </div>
      <div class="message">
        <strong>
          {title}
        </strong>
        {description}
      </div>
      <div>
        <button on:click={closeAlert}>&#x2716;</button>
      </div>
    </dialog>
  {/if}
</div>

<style>
  dialog {
    min-width: 300px;
    display: flex;
    justify-content: space-between;
    font-family: Arial, Helvetica, sans-serif;
    border: none;
  }

  button {
    background: none;
    border: none;
    font-size: 21px;
  }

  .icon {
    margin-right: 10px;
  }

  .message {
    display: flex;
    flex-direction: column;
    line-height: 24px;
    min-width: 300px;
  }

  .fade-in {
    animation: fade-in 2000ms both;
  }

  @keyframes fade-in {
    from {
      opacity: 0%;
    }
  }

  .alert-warn {
    background: #ffeb3b;
    color: #000000;
  }

  .alert-info {
    background: #2196f3;
    color: #ffffff;
  }

  .alert-error {
    background: #ff0000;
    color: #ffffff;
  }
</style>
