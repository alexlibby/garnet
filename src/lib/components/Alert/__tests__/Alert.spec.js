import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Alert from "../../Alert/Alert.svelte";

const AlertHTML = `<div class="garnet s-4Y75ZjC11XWf"><dialog class="alert alert-info fade-in s-4Y75ZjC11XWf" role="alertdialog"  title="Alert Dialog"><div class="icon s-4Y75ZjC11XWf"><svg width="24px" height="24px" viewBox="0 0 512 512"><path d="M248 64C146.39 64 64 146.39 64 248s82.39 184 184 184 184-82.39 184-184S349.61 64 248 64z" fill="none" stroke="currentColor" stroke-miterlimit="10" stroke-width="32"></path><path fill="none" stroke="#ffffff" stroke-linecap="round" stroke-linejoin="round" stroke-width="32" d="M220 220h32v116"></path><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="32" d="M208 340h88"></path><path d="M248 130a26 26 0 1026 26 26 26 0 00-26-26z" fill="#ffffff"></path></svg><!--<Icon>--></div> <div class="message s-4Y75ZjC11XWf"><strong class="s-4Y75ZjC11XWf">Simple Info</strong> An info description</div> <div class="s-4Y75ZjC11XWf"><button class="s-4Y75ZjC11XWf">âœ–</button></div></dialog></div>`;
expect.extend(AxeMatchers);

describe("Tests for Alert", () => {
  const props = {
    show: true,
    description: "An info description",
    title: "Simple Info",
    showIcon: "true",
    type: "info",
  };

  test("should render properly", async () => {
    const result = await render(Alert, { props });
    expect(() => result).not.toThrow();
  });

  // https://testing-library.com/docs/svelte-testing-library/example
  test("should disappear when close button clicked", async () => {
    const result = await render(Alert, { props });
    const alertDialog = fireEvent.click(screen.getByRole("alertdialog"));
    expect(alertDialog).toBeNull;
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Alert({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => AlertHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
