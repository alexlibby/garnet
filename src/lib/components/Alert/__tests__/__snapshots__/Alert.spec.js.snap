// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Alert > get a snapshot of component 1`] = `
"<div class="garnet s-4Y75ZjC11XWf"><dialog class="alert alert-info fade-in s-4Y75ZjC11XWf" role="alertdialog" title="Alert Dialog"><div class="icon s-4Y75ZjC11XWf"><svg width="24px" height="24px" viewBox="0 0 512 512"><path d="M248 64C146.39 64 64 146.39 64 248s82.39 184 184 184 184-82.39 184-184S349.61 64 248 64z" fill="none" stroke="currentColor" stroke-miterlimit="10" stroke-width="32"></path><path fill="none" stroke="#ffffff" stroke-linecap="round" stroke-linejoin="round" stroke-width="32" d="M220 220h32v116"></path>
      <path fill="none" stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="32" d="M208 340h88"></path><path d="M248 130a26 26 0 1026 26 26 26 0 00-26-26z" fill="#ffffff"></path></svg><!--<Icon>--></div> <div class="message s-4Y75ZjC11XWf"><strong class="s-4Y75ZjC11XWf">Simple Info</strong> An info description</div> <div class="s-4Y75ZjC11XWf"><button class="s-4Y75ZjC11XWf">âœ–</button></div></dialog></div><!--<Alert>-->"
`;
